### PR TITLE
feat: add webhook

### LINKED ISSUE #8: Generate new diffs when a new version is published
We need to generate diffs whenever a new version is posted (and update the static paths after we're done with it).
There are 2 ways I can think of to do it:
1. Have a cron job that runs weekly to check if there's a new version and then generate all diffs
2. Ask t3-oss team for a webhook for releases and create an endpoint in this app to expect that webhook and generate all diffs.

### PR DESCRIPTION
**Pull request by PiotrekPKP**
Adds webhook to the project (closes #8)

### Important
- you have to set the `WEBHOOK_SECRET` env variable to whatever is set in `create-t3-app` repo
- you can run `webhookthing` using `npm run webhook:dev` and then run the webhook from there, BUT you'll have to have the sha256 signature that github sends you and use it with `WEBHOOK_SIGNATURE` env variable (maybe there's a better way to do it?)
- I couldn't test the webhook properly without `create-t3-app` repo

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,14 +1,2 @@
...
-# ".env" and populate it with your secrets.
-
-# When adding additional environment variables, the schema in "/src/env.mjs"
-# should be updated accordingly.
-
-# Example:
-# SERVERVAR="foo"
-# NEXT_PUBLIC_CLIENTVAR="bar"
+# GitHub webhook secret
+WEBHOOK_SECRET=
```
Comment (andreifilip123) [inline .env.example]:
```suggestion
WEBHOOK_SECRET=

```

File: .thing/hooks/new-release.config.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "url": "http://localhost:3000/api/webhook",
+  "headers": {
+    "Content-Type": "application/json",
+    "X-Hub-Signature-256": "%%WEBHOOK_SIGNATURE%%"
+  }
+}
```
Comment (andreifilip123) [inline .thing/hooks/new-release.config.json]: Add empty lines at end of files

File: src/env.mjs [hunk 3]
```diff
@@ -6,6 +6,7 @@ import { z } from "zod";
  */
 const server = z.object({
   NODE_ENV: z.enum(["development", "test", "production"]),
+  WEBHOOK_SECRET: z.string().min(1),
```
Comment (andreifilip123) [inline src/env.mjs:9]: Do you need the `.min(1)`? Doesn't the lack of `.optional()`just do the trick ?

File: package.json [hunk 4]
```diff
@@ -6,7 +6,8 @@
     "build": "next build",
     "dev": "next dev",
     "lint": "next lint",
-    "start": "next start"
+    "start": "next start",
+    "webhook:dev": "WEBHOOK_SIGNATURE=XXX npx webhookthing@latest"
```
Comment (andreifilip123) [inline package.json]: Does the env var need to be here ?

File: src/pages/api/webhook.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  if (event.data.action !== "released") {
+    //Ignore other events, we will just use the "released" event
+    return res.status(200).json({ success: true });
+  }
+
+  res.status(200).json({
+    success: true,
+  });
```
Comment (andreifilip123) [inline src/pages/api/webhook.ts:44]: Doesn't this do early return ?

File: src/env.mjs (hunk 3 above)
Comment (PiotrekPKP) [reply src/env.mjs:9]: No, it accepts `""` without the `.min(1)`

File: package.json (hunk 4 above)
Comment (PiotrekPKP) [reply package.json]: I already removed it :)

File: src/pages/api/webhook.ts (hunk 5 above)
Comment (PiotrekPKP) [reply src/pages/api/webhook.ts:44]: Nope, tested it. It works in the background but sends a response so GitHub doesn't terminate the webhook.

Comment (andreifilip123) [reply src/pages/api/webhook.ts:44]: Does GitHub care for the response ?

Comment (PiotrekPKP) [reply src/pages/api/webhook.ts:44]: Yep - without a response in 10 seconds it terminates the `POST` request and kills it.

Comment (PiotrekPKP) [reply src/pages/api/webhook.ts:44]: https://docs.github.com/en/rest/guides/best-practices-for-integrators?apiVersion=2022-11-28#favor-asynchronous-work-over-synchronous

### Quote:
GitHub expects that integrations respond within 10 seconds of receiving the webhook payload. If your service takes longer than that to complete, then GitHub terminates the connection and the payload is lost.

**Comment by PiotrekPKP** [comment]
In the latest commit I've also enabled revalidation for changed pages in the webhook, forgot to do that earlier :)

**Comment by andreifilip123** [comment]
What env var do you need me to add to Vercel and where do I get it from ?

**Comment by PiotrekPKP** [comment]
> What env var do you need me to add to Vercel and where do I get it from ?

`WEBHOOK_SECRET` - Nexxel has to set it in webhook settings and send it to you / you have to generate one and send it to Nexxel

**Comment by andreifilip123** [comment]
Thank you @PiotrekPKP! Finished the PR in #24 (couldn't push to this directly). Appreciate your help very much!

**Pull request closed without merging**